### ISSUE TITLE
Generic.WhiteSpace.DisallowSpaceIndent.SpacesUsed shouldn't show waring if used in an statement

### ISSUE THREAD

**Issue by JoryHogeveen** [squizlabs/PHP_CodeSniffer#1586]
When using multiline `if` statements etc. this check generates an error when formatted like this:

```php
	if ( CONDITION &&
	     CONDITION &&
	     CONDITION
	) {
		//...
	}
```
The whole block is indented with tabs but within the if statement shouldn't it be indented as spaces for alignment dispite your tabs width settings?

**Comment by gsherwood** [comment]
That code doesn't generate any errors for me, but I am assuming you've indented the 2nd and 3rd condition lines with tabs and then a single space to align the condition text. Like this:
```
if·($foo)·{
»   if·(·CONDITION·&&
»   »   ·CONDITION·&&
»   »   ·CONDITION
»   )·{
»   »   //...
»   }
}
```

If you've instead done something like this:
```
if·($foo)·{
»   if·(·CONDITION·&&
»   ·····CONDITION·&&
»   ·····CONDITION
»   )·{
»   »   //...
»   }
}
```

Then the sniff is going to produce errors because some of those spaces can be replaced by a single tab.

If you can tell me exactly how your code is formatted, I might be able to help figure out if there is a bug in the sniff.

**Comment by JoryHogeveen** [comment]
Hi @gsherwood ,

I indeed indent like your second example. Using tabs to for the indentation of the full `if` block and spaces if after that.
I also think this _should_ be the correct way since it allows for fixed alignment of conditions. If you use tabs the alignment will change when you change the tab size.

For example (using the example that is correct according to the current standards):

```
if·($foo)·{
»   if·(·CONDITION·&&
»       ·CONDITION·&&
»       ·CONDITION
»   )·{
»   »   //...
»   }
}
```

When you change the tab size to 8 (like GitHub does) it would look like this:
```
if·($foo)·{
»       if·(·CONDITION·&&
»               ·CONDITION·&&
»               ·CONDITION
»       )·{
»       »       //...
»       }
}
```

**Comment by JoryHogeveen** [comment]
Oh, same thing for long string alignments:

Not the the last `.` is actual PHP.
```
»   $var = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut sagittis, '
»   ........ 'neque a venenatis gravida, tellus nisi dictum justo, ut euismod nisi justo placerat libero. '
»   ........ 'Curabitur vitae sapien id lacus auctor vehicula eu eu quam.';
```
This would align perfectly, no matter the tab size while the codeblock below doesnt:
```
»   $var = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut sagittis, '
»   »   .... 'neque a venenatis gravida, tellus nisi dictum justo, ut euismod nisi justo placerat libero. '
»   »   .... 'Curabitur vitae sapien id lacus auctor vehicula eu eu quam.';
```
On GitHub this would look like:
```
»       $var = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut sagittis, '
»       »       .... 'neque a venenatis gravida, tellus nisi dictum justo, ut euismod nisi justo placerat libero. '
»       »       .... 'Curabitur vitae sapien id lacus auctor vehicula eu eu quam.';
```

**Comment by JoryHogeveen** [comment]
One last thing:

> Then the sniff is going to produce errors because some of those spaces can be replaced by a single tab.

So this means that the sniff is fixed on a tab size of 4 spaces?

**Comment by jrfnl** [comment]
> So this means that the sniff is fixed on a tab size of 4 spaces?

No, you can set the tab width yourself either via the command line `--tab-width=4` or via a custom ruleset: `<arg name="tab-width" value="4"/>`.

**Comment by JoryHogeveen** [comment]
Ah ok, thanks.
But still. Isn't the whole purpose of using tabs properly to allow other users to set their own tab size without messing up the indentation alignment?

**Comment by jrfnl** [comment]
There's a difference between indentation and alignment.
Indentation can be done with both tabs as well as spaces. Alignment, especially precision alignment only with spaces. Based on your previous comments, you seem to be looking for precision alignment and that's not something that can be done with tabs, let alone be auto-fixed.

**Comment by JoryHogeveen** [comment]
Yep thats my point :)
I wouldn't say that precision alignment should be a warning with this sniff, only incorrect usage of spaces.
Don't know what you mean with auto-fixed though.

**Comment by gsherwood** [comment]
Point taken about tab widths changing the precision alignment. The sniff would need to be modified to support these specific cases.

**Comment by jrfnl** [comment]
> The sniff would need to be modified to support these specific cases.

Can it though ? The op wants the sniff to ignore any precision alignment within function call parenthesis. Someone else may want them to be ignored within array parenthesis. So how could the sniff ever distinguish all those different cases and still do the normal spaces to tabs conversion  for the rest of us ?

**Comment by JoryHogeveen** [comment]
Anything can, but it will be difficult I guess.

> Someone else may want them to be ignored within array parenthesis.

Array are always indented with tabs, it's about multiline blocks for `if`/`elseif`/`while` etc. and multiline strings (separated with `.`).
However indeed, multiline strings within an array block should be valid:
```
»   $arr = array(
»   »   'key'  => 1,
»   »   'key2' => 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut sagittis, '
»   »   ......... . 'neque a venenatis gravida, tellus nisi dictum justo, ut euismod nisi justo placerat libero. '
»   »   ......... . 'Curabitur vitae sapien id lacus auctor vehicula eu eu quam.';
»   );
```

I guess one option is to make specific exceptions for these types in the sniff, no idea how though.
Also, maybe make it optionally? `<arg name="allow-precision-alignment" value="1"/>.` or something?

**Comment by jrfnl** [comment]
@JoryHogeveen My point is that any choices made here are dependent on the rules set by a project. So the "specific exceptions" are project based exceptions, not industry rules or even common practice.

**Comment by JoryHogeveen** [comment]
@jrfnl  That is why I thought it would be better to make it optional. This way the user can define wether they would like to use precision alignment or not.
phpStorm for example has it build in by default for codeformatting and you can easily enable/disable it. Same goes for Zend Studio.

**Comment by gsherwood** [comment]
Yes, this would have to be optional. I think it's probably achievable in specific situations. I'm not sure if it is achievable in all situations.
